### PR TITLE
address_module

### PR DESCRIPTION
**Pull request by nikpawar89**
Squashed multiple commits to single commit.

### CODE REVIEW

File: fineract-provider/src/main/java/org/apache/fineract/commands/domain/CommandSource.java [hunk 1]
```diff
@@ -95,6 +95,7 @@
```
Comment (Nayan) [inline fineract-provider/src/main/java/org/apache/fineract/commands/domain/CommandSource.java:95]: @nikpawar89  this file should be ignored, has only formatting changes

File: fineract-provider/src/main/resources/sql/migrations/core_db/v312_address_module_tables_metadat.sql.bak [hunk 2]
```diff
@@ -0,0 +1,96 @@
+--
```
Comment (Nayan) [inline fineract-provider/src/main/resources/sql/migrations/core_db/v312_address_module_tables_metadat.sql.bak]: @nikpawar89 why backup/temp file is part of this commit 👎

File: fineract-provider/src/main/java/org/apache/fineract/commands/domain/CommandWrapper.java [hunk 3]
```diff
@@ -38,6 +38,8 @@
     private final String json;
     private final String transactionId;
     private final Long productId;
+    private final Boolean status;
```
Comment (Nayan) [inline fineract-provider/src/main/java/org/apache/fineract/commands/domain/CommandWrapper.java:41]: @nikpawar89  what is the use this status field? avoid modifying command handler infrastructure code

Comment (nikpawar89) [inline fineract-provider/src/main/java/org/apache/fineract/commands/domain/CommandWrapper.java:41]: THe status field is basically, the status of the address. 
I could not find any better way to pass the field( which is a command) from API resource to command handler.

File: fineract-provider/src/main/resources/sql/migrations/core_db/v312_address_module_tables_metadat.sql.bak (hunk 2 above)
Comment (nikpawar89) [inline fineract-provider/src/main/resources/sql/migrations/core_db/v312_address_module_tables_metadat.sql.bak]: would remove that

File: fineract-provider/src/main/java/org/apache/fineract/commands/domain/CommandSource.java (hunk 1 above)
Comment (nikpawar89) [inline fineract-provider/src/main/java/org/apache/fineract/commands/domain/CommandSource.java:95]: removed the file from commit

File: fineract-provider/src/main/java/org/apache/fineract/commands/service/CommandWrapperBuilder.java [hunk 4]
```diff
@@ -41,10 +41,39 @@
...
-                this.entityName, this.entityId, this.subentityId, this.href, this.json, this.transactionId, this.productId, this.templateId);
+                this.entityName, this.entityId, this.subentityId, this.href, this.json, this.transactionId, this.productId, this.templateId
+                ,this.status);
+    }
+    
+    public CommandWrapperBuilder addClientAddress(final long clientId,final long addressTypeId) {
+        this.actionName = "CREATE";
+        this.entityName = "ADDRESS";
+        this.entityId = addressTypeId;
+        this.href = "/Address/template";
```
Comment (nazeer1100126) [inline fineract-provider/src/main/java/org/apache/fineract/commands/service/CommandWrapperBuilder.java:56]: href value should reflect actual path url

File: fineract-provider/src/main/java/org/apache/fineract/commands/service/CommandWrapperBuilder.java [hunk 5]
```diff
@@ -41,10 +41,39 @@
...
+        this.clientId=clientId;
+        return this;
+    }
+    
+    public CommandWrapperBuilder updateClientAddress(final long clientId,final long addressTypeID,final boolean status) {
+        this.actionName = "UPDATE";
+        this.entityName = "ADDRESS";
+        this.entityId = addressTypeID;
+        this.href = "/Address/template";
```
Comment (nazeer1100126) [inline fineract-provider/src/main/java/org/apache/fineract/commands/service/CommandWrapperBuilder.java:65]: href value should reflect actual path url

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/address/api/AddressConfigurationApiResources.java [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+		this.toApiJsonSerializerfld = toApiJsonSerializerfld;
+		this.apiRequestParameterHelper = apiRequestParameterHelper;
+		this.commandsSourceWritePlatformService = commandsSourceWritePlatformService;
+	}
+
+	@GET
+	@Path("/fieldconfiguration/{entity}")
+	@Consumes({ MediaType.APPLICATION_JSON })
+	@Produces({ MediaType.APPLICATION_JSON })
+	public String getAddresses(@PathParam("entity") final String entityname, @Context final UriInfo uriInfo) {
```
Comment (nazeer1100126) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/address/api/AddressConfigurationApiResources.java]: Already EntityFieldConfigurationApiResources.getAddresses(...) API is returning same Data. Why this API is required in this class?

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/address/domain/Address.java [hunk 7]
```diff
@@ -0,0 +1,343 @@
...
+import org.springframework.data.jpa.domain.AbstractPersistable;
+
+import com.google.gson.JsonObject;
+
+@Entity
+@Table(name = "m_address")
+public class Address extends AbstractPersistable<Long> {
+
+	@OneToMany(mappedBy = "address", cascade = CascadeType.ALL)
+	private List<ClientAddress> clientaddress = new ArrayList<>();
```
Comment (nazeer1100126) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/address/domain/Address.java]: Collection child entities should be of type Set. 
Reason: The moment we add more child collections, we end up having lot of issues like Multiple param bags, duplicate entries in list child entity.

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/address/service/AddressWritePlatformServiceImpl.java [hunk 8]
```diff
@@ -0,0 +1,317 @@
...
+		CodeValue countryIdObj;
+
+		boolean is_address_update = false;
+
+		boolean is_address_config_update = false;
+
+		final CodeValue addresstyp = this.codeValueRepository.getOne(addressTypeId);
+		
+		
+		System.out.println("clientId is"+ clientId);
```
Comment (nazeer1100126) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/address/service/AddressWritePlatformServiceImpl.java]: Remove sysouts.

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/address/service/AddressWritePlatformServiceImpl.java [hunk 9]
```diff
@@ -0,0 +1,317 @@
...
+		System.out.println("addresstyp:" +addresstyp);
+		System.out.println("status:" +status);
+
+		final ClientAddress clientAddressObj = this.clientAddressRepositoryWrapper
+				.findOneByClientIdAndAddressTypeAndIsActive(clientId, addresstyp, status);
+		System.out.println(clientAddressObj);
```
Comment (nazeer1100126) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/address/service/AddressWritePlatformServiceImpl.java]: Remove sysout

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/address/service/AddressWritePlatformServiceImpl.java [hunk 10]
```diff
@@ -0,0 +1,317 @@
...
+		boolean is_address_config_update = false;
+
+		final CodeValue addresstyp = this.codeValueRepository.getOne(addressTypeId);
+		
+		
+		System.out.println("clientId is"+ clientId);
+		System.out.println("addresstyp:" +addresstyp);
+		System.out.println("status:" +status);
+
+		final ClientAddress clientAddressObj = this.clientAddressRepositoryWrapper
```
Comment (nazeer1100126) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/address/service/AddressWritePlatformServiceImpl.java:166]: Why this method is required to pass status? Any how you are always retrieving active client address right?

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/client/domain/ClientAddressRepositoryWrapper.java [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+@Service
+public class ClientAddressRepositoryWrapper {
+	private final ClientAddressRepository clientAddressRepository;
+
+	@Autowired
+	public ClientAddressRepositoryWrapper(final ClientAddressRepository clientAddressRepository) {
+		this.clientAddressRepository = clientAddressRepository;
+	}
+
+	public ClientAddress findOneByClientIdAndAddressTypeAndIsActive(final long clientId, final CodeValue addressType,
```
Comment (nazeer1100126) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/client/domain/ClientAddressRepositoryWrapper.java:34]: This method returns active client addresses. is it required to have isActive as param?

File: fineract-provider/src/main/java/org/apache/fineract/commands/domain/CommandWrapper.java (hunk 3 above)
Comment (nazeer1100126) [inline fineract-provider/src/main/java/org/apache/fineract/commands/domain/CommandWrapper.java:41]: No need to add status variable

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/shareproducts/service/ShareProductCommandsServiceImpl.java [hunk 12]
```diff
@@ -47,7 +47,7 @@ public CommandProcessingResult postDividends(Long productId, JsonCommand jsonCom
     public Object handleCommand(Long productId, String command, String jsonBody) {
         final JsonElement parsedCommand = this.fromApiJsonHelper.parse(jsonBody);
         final JsonCommand jsonCommand = JsonCommand.from(jsonBody, parsedCommand, this.fromApiJsonHelper, null, null, null, null, null,
-                null, null, null, null, null);
+                null, null, null, null, null,null);
```
Comment (nazeer1100126) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/shareproducts/service/ShareProductCommandsServiceImpl.java:50]: No change is required in this class.

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/shareaccounts/service/ShareAccountCommandsServiceImpl.java [hunk 13]
```diff
@@ -47,7 +47,7 @@ public ShareAccountCommandsServiceImpl(final FromJsonHelper fromApiJsonHelper,
     public Object handleCommand(Long accountId, String command, String jsonBody) {
         final JsonElement parsedCommand = this.fromApiJsonHelper.parse(jsonBody);
         final JsonCommand jsonCommand = JsonCommand.from(jsonBody, parsedCommand, this.fromApiJsonHelper, null, null, null, null, null,
-                null, null, null, null, null);
+                null, null, null, null, null,null);
```
Comment (nazeer1100126) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/shareaccounts/service/ShareAccountCommandsServiceImpl.java:50]: No change is required in this class

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanWritePlatformServiceJpaRepositoryImpl.java [hunk 14]
```diff
@@ -2292,7 +2292,7 @@ public void applyOverdueChargesForLoan(final Long loanId, Collection<OverdueLoan
 
             final JsonElement parsedCommand = this.fromApiJsonHelper.parse(overdueInstallment.toString());
             final JsonCommand command = JsonCommand.from(overdueInstallment.toString(), parsedCommand, this.fromApiJsonHelper, null, null,
-                    null, null, null, loanId, null, null, null, null);
+                    null, null, null, loanId, null, null, null, null,null);
```
Comment (nazeer1100126) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanWritePlatformServiceJpaRepositoryImpl.java:2295]: No change is required in this class

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/client/service/ClientWritePlatformServiceJpaRepositoryImpl.java [hunk 15]
```diff
@@ -195,8 +203,138 @@ private void handleDataIntegrityIssues(final JsonCommand command, final DataInte
     @Transactional
     @Override
     public CommandProcessingResult createClient(final JsonCommand command) {
-
-        try {
+    	 try {
+             final AppUser currentUser = this.context.authenticatedUser();
+             
+             final long configId=29;
```
Comment (nazeer1100126) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/client/service/ClientWritePlatformServiceJpaRepositoryImpl.java:209]: Why this config is hard coded here?

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/client/service/ClientWritePlatformServiceJpaRepositoryImpl.java [hunk 16]
```diff
@@ -195,8 +203,138 @@ private void handleDataIntegrityIssues(final JsonCommand command, final DataInte
...
+             CommandProcessingResult result = openSavingsAccount(newClient, fmt);
+             if (result.getSavingsId() != null) {
+                 this.clientRepository.save(newClient);
+             }
+             
+             if(isEntity)            
+             	extractAndCreateClientNonPerson(newClient, command);
+             
+             final long clientId=newClient.getId();
+             System.out.println("client with client ID created:"+clientId);
```
Comment (nazeer1100126) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/client/service/ClientWritePlatformServiceJpaRepositoryImpl.java]: remove sysout

File: fineract-provider/src/main/resources/sql/migrations/core_db/v312_address_module_tables_metadat.sql [hunk 17]
```diff
@@ -0,0 +1,127 @@
...
+	`updated_on` DATE NULL DEFAULT NULL,
+	PRIMARY KEY (`id`),
+	INDEX `address_fields_codefk1` (`state_province_id`),
+	INDEX `address_fields_codefk2` (`country_id`),
+	CONSTRAINT `address_fields_codefk1` FOREIGN KEY (`state_province_id`) REFERENCES `m_code_value` (`id`),
+	CONSTRAINT `address_fields_codefk2` FOREIGN KEY (`country_id`) REFERENCES `m_code_value` (`id`)
+)
+COLLATE='utf8_general_ci'
+ENGINE=InnoDB
+AUTO_INCREMENT=20
```
Comment (nazeer1100126) [inline fineract-provider/src/main/resources/sql/migrations/core_db/v312_address_module_tables_metadat.sql]: AUTO_INCREMENT=20 ?

File: fineract-provider/src/main/resources/sql/migrations/core_db/v312_address_module_tables_metadat.sql [hunk 18]
```diff
@@ -0,0 +1,127 @@
...
+	PRIMARY KEY (`id`),
+	INDEX `addressIdFk` (`address_id`),
+	INDEX `address_codefk` (`address_type_id`),
+	INDEX `clientaddressfk` (`client_id`),
+	CONSTRAINT `address_codefk` FOREIGN KEY (`address_type_id`) REFERENCES `m_code_value` (`id`),
+	CONSTRAINT `clientaddressfk` FOREIGN KEY (`client_id`) REFERENCES `m_client` (`id`)
+)
+COLLATE='utf8_general_ci'
+ENGINE=InnoDB
+AUTO_INCREMENT=28
```
Comment (nazeer1100126) [inline fineract-provider/src/main/resources/sql/migrations/core_db/v312_address_module_tables_metadat.sql]: why AUTO_INCREMENT=28 ?

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/client/handler/UpdateClientAddressCommandHandler.java [hunk 19]
```diff
@@ -0,0 +1,47 @@
...
+
+	@Autowired
+	public UpdateClientAddressCommandHandler(final AddressWritePlatformService writePlatformService) {
+		this.writePlatformService = writePlatformService;
+	}
+
+	@Override
+	public CommandProcessingResult processCommand(final JsonCommand command) {
+		return this.writePlatformService.updateClientAddress(command.getClientId(), command.entityId(),
+				command.getStatus(), command);
```
Comment (nazeer1100126) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/client/handler/UpdateClientAddressCommandHandler.java:43]: status is not required to pass

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/address/api/AddressConfigurationApiResources.java (hunk 6 above)
Comment (nikpawar89) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/address/api/AddressConfigurationApiResources.java]: removed the redundant method from AddressConfigurationApiResources.java

File: fineract-provider/src/main/resources/sql/migrations/core_db/v312_address_module_tables_metadat.sql (hunk 18 above)
Comment (nikpawar89) [inline fineract-provider/src/main/resources/sql/migrations/core_db/v312_address_module_tables_metadat.sql]: changed to 1

File: fineract-provider/src/main/resources/sql/migrations/core_db/v312_address_module_tables_metadat.sql (hunk 17 above)
Comment (nikpawar89) [inline fineract-provider/src/main/resources/sql/migrations/core_db/v312_address_module_tables_metadat.sql]: changed to 1

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/client/service/ClientWritePlatformServiceJpaRepositoryImpl.java (hunk 16 above)
Comment (nikpawar89) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/client/service/ClientWritePlatformServiceJpaRepositoryImpl.java]: removed

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/client/service/ClientWritePlatformServiceJpaRepositoryImpl.java (hunk 15 above)
Comment (nikpawar89) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/client/service/ClientWritePlatformServiceJpaRepositoryImpl.java:209]: I am giving insert for this specific configuration and hence I have hardcoded it.

Comment (nazeer1100126) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/client/service/ClientWritePlatformServiceJpaRepositoryImpl.java:209]: What happens if c_configuration already having a row with id 29?

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/client/domain/ClientAddressRepositoryWrapper.java (hunk 11 above)
Comment (nikpawar89) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/client/domain/ClientAddressRepositoryWrapper.java:34]: isActive is a parameter which helps retrieving both active and inactive address.

File: fineract-provider/src/main/java/org/apache/fineract/commands/service/CommandWrapperBuilder.java (hunk 4 above)
Comment (nikpawar89) [inline fineract-provider/src/main/java/org/apache/fineract/commands/service/CommandWrapperBuilder.java:56]: I actually don't require them. Can I omit these entries?

File: fineract-provider/src/main/java/org/apache/fineract/commands/service/CommandWrapperBuilder.java (hunk 5 above)
Comment (nikpawar89) [inline fineract-provider/src/main/java/org/apache/fineract/commands/service/CommandWrapperBuilder.java:65]: I actually don't require them. Can I omit these entries?

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/address/service/AddressWritePlatformServiceImpl.java (hunk 9 above)
Comment (nikpawar89) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/address/service/AddressWritePlatformServiceImpl.java]: done

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/address/service/AddressWritePlatformServiceImpl.java (hunk 8 above)
Comment (nikpawar89) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/address/service/AddressWritePlatformServiceImpl.java]: done

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/address/service/AddressWritePlatformServiceImpl.java (hunk 10 above)
Comment (nikpawar89) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/address/service/AddressWritePlatformServiceImpl.java:166]: we are retrieving both active and non-active address, hence a filter parameter is required.
The status parameter is mainly required when you are toggling from active -inactive address.

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/shareproducts/service/ShareProductCommandsServiceImpl.java (hunk 12 above)
Comment (nikpawar89) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/shareproducts/service/ShareProductCommandsServiceImpl.java:50]: from static method of JsonCommand class has to be chaned in order to get status value from the JSONCommand.  The change in the signature results in addition of one more parameter value in this class.

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/shareaccounts/service/ShareAccountCommandsServiceImpl.java (hunk 13 above)
Comment (nikpawar89) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/shareaccounts/service/ShareAccountCommandsServiceImpl.java:50]: from static method of JsonCommand class has to be chaned in order to get status value from the JSONCommand.  The change in the signature results in addition of one more parameter value in this class.

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanWritePlatformServiceJpaRepositoryImpl.java (hunk 14 above)
Comment (nikpawar89) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanWritePlatformServiceJpaRepositoryImpl.java:2295]: from static method of JsonCommand class has to be chaned in order to get status value from the JSONCommand.  The change in the signature results in addition of one more parameter value in this class.

File: fineract-provider/src/main/resources/sql/migrations/core_db/v312_address_module_tables_metadat.sql [hunk 20]
```diff
@@ -0,0 +1,127 @@
...
+--code inserts
+INSERT INTO `m_code` (`id`, `code_name`, `is_system_defined`) VALUES (150, 'INDIAN STATES', 0);
+INSERT INTO `m_code` (`id`, `code_name`, `is_system_defined`) VALUES (151, 'COUNTRY ID', 0);
+INSERT INTO `m_code` (`id`, `code_name`, `is_system_defined`) VALUES (152, 'ADDRESS TYPE ID', 0);
+
+--add new column in portfolio
+alter table `m_portfolio_command_source` add column `status` TINYINT(4) NULL DEFAULT NULL;
+
+--configuration
+INSERT INTO `c_configuration` (`id`, `name`, `value`, `date_value`, `enabled`, `is_trap_door`, `description`) VALUES (29, 'Enable-Address', NULL, NULL, 1, 0, NULL);
```
Comment (Nayan) [inline fineract-provider/src/main/resources/sql/migrations/core_db/v312_address_module_tables_metadat.sql:29]: @nikpawar89 primary key should be auto generated

File: fineract-provider/src/main/resources/sql/migrations/core_db/v312_address_module_tables_metadat.sql [hunk 21]
```diff
@@ -0,0 +1,127 @@
...
+;
+
+
+/*!40101 SET @OLD_CHARACTER_SET_CLIENT=@@CHARACTER_SET_CLIENT */;
+/*!40101 SET NAMES utf8mb4 */;
+/*!40014 SET @OLD_FOREIGN_KEY_CHECKS=@@FOREIGN_KEY_CHECKS, FOREIGN_KEY_CHECKS=0 */;
+/*!40101 SET @OLD_SQL_MODE=@@SQL_MODE, SQL_MODE='NO_AUTO_VALUE_ON_ZERO' */;
+-- Dumping data for table mifostenant-default.m_field_configuration: ~18 rows (approximately)
+/*!40000 ALTER TABLE `m_field_configuration` DISABLE KEYS */;
+INSERT INTO `m_field_configuration` (`id`, `entity`, `table`, `field`, `is_enabled`, `is_mandatory`, `validation_regex`) VALUES
```
Comment (Nayan) [inline fineract-provider/src/main/resources/sql/migrations/core_db/v312_address_module_tables_metadat.sql:105]: @nikpawar89 primary key should be auto generated

**Pull request closed without merging**